### PR TITLE
Finalize user management, add CSRF token protection and styling.

### PR DESCRIPTION
**Pull request by cmeiklejohn**
This squashed commit obsoletes #15.
- CSRF resource protection.
- Serve up application template as DTL.
- Add loading spinner.
- Style creation form.
- Highlight admin rows.
- Disable "disable/revoke" buttons for admins.
- Move all work to the gen_server.
- Refactor module concerns.
- Run dialyzer and update Makefile.
- Admin key in user responses.

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
Newman did most of the JavaScript and styling, so I'm +1 on that.  Just need a review of the Erlang, @kellymclaughlin .

**Comment by reiddraper** [comment]
Not sure it actually matters, but for consistency, we should be using 'new-style' specs for new Erlang. So

``` erlang
%% @spec configure_s3_connection() -> term()
```

becomes

``` erlang
-spec configure_s3_connection() -> term().
```

**Comment by reiddraper** [comment]
Getting this error running `make dialyzer` (after building plt)

```
dialyzer: Analysis failed with error:
Could not scan the following file(s): [{"/Users/reid/Documents/repos/basho/riak_cs_control/ebin/application_dtl.beam",
                                        "  Could not get abstract code for: /Users/reid/Documents/repos/basho/riak_cs_control/ebin/application_dtl.beam\n  Recompile with +debug_info or analyze starting from source code"}]
```

**Comment by cmeiklejohn** [comment]
That's a bug I've filed against rebar -- it doesn't build erlydtl templates with debug symbols.

**Comment by cmeiklejohn** [comment]
@reiddraper pushed an update to convert everything to new-style specs.

File: src/riak_cs_control_security.erl [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+%% @doc Helpers for security of requests.
+
+-module(riak_cs_control_security).
+-author('Christopher Meiklejohn <cmeiklejohn@basho.com>').
+
+-export([csrf_token/2,
+         is_protected/2]).
+
+%% @doc Store a CSRF protection token in a cookie.
```
Comment (reiddraper) [inline src/riak_cs_control_security.erl]: should this be s/Store/Get/ ?

File: src/riak_cs_control_wm_asset.erl [hunk 2]
```diff
@@ -21,39 +21,86 @@
...
+%% @doc Given a series of request tokens, normalize to priv dir file.
+-spec normalize_filepath(list()) -> list().
+normalize_filepath(Filepath) ->
+    {ok, App} = application:get_application(?MODULE),
+    filename:join([code:priv_dir(App), 'www'] ++ Filepath).
 
+%% @doc Return a context which determines if we serve up the application
+%% template or a file resource.
+identify_resource(ReqData, Context) ->
+    case Context#context.resource of
```
Comment (reiddraper) [inline src/riak_cs_control_wm_asset.erl]: maybe replace this nested case with smaller funs?

**Comment by reiddraper** [comment]
The `rebar.config` uses `erlydtl` master, which appears to cause this error:

```
17:17:20.563 [error] webmachine error: path="/"{error,{error,undef,[{erlydtl_runtime,find_value,[csrf_token,[{csrf_token,"tOqS0_0CcNx3nqd1ClydkhqM-Tv4e6wm_JzyYWBdCt-24UghhZKacAMOnWKaXIhNQjl4KnLmFO8sgfwy8FHZK8ER9K5_6VASID743qVMsuLL2NgEKPFc1uWHa-otNF3YEE_7BDGuUiqdIk4QL7tcu4WWLIORtVfEkN7R9zo4iexPA27fwqFqW4I1OV-MAccZp17rH
-CSFg4UuUv67dBSN5URA02QSHAsCGkzHOaQMsyTyVutAupBmLGX5QX59AQl2AaNcT3S_g4uqtlh3UgAwRj0cyioi7TJrfYpQcIIKz7UzeiJgTZaByptoP9ZQeOUjVgX9Vs7VgBwBZpRdaEseg"}]],[]},{application_dtl,render_internal,4,[]},{application_dtl,render,2,[]},{riak_cs_control_wm_asset,to_resource,2,[{file,"s
rc/riak_cs_control_wm_asset.erl"},{line,97}]},{webmachine_resource,resource_call,3,[{file,"src/webmachine_resource.erl"},{line,171}]},{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,130}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmach
ine_decision_core.erl"},{line,48}]},{webmachine_decision_core,decision,1,[{file,"src/webmachine_decision_core.erl"},{line,555}]}]}}
```

master for me is currently `3306858bbceab3d4b2c792e69aa4d454d5fd05f5`. Checking out `6d1fcaf7edcfb92e23e1d1e86b0702e93928962a` fixed the issue. I did have to re-build the whole rel, I couldn't just 'make' again, since I had done 'make stage'.

**Comment by reiddraper** [comment]
Not sure we care, but if I send the wrong csrf token, the API appropriately returns HTTP 403, but the GUI still updates as if the action succeeded.

File: src/riak_cs_control_security.erl (hunk 1 above)
Comment (cmeiklejohn) [inline src/riak_cs_control_security.erl]: Addressed in [0bc91097d6c79cf9bd79ae3a2012094a7d65f13b].

File: src/riak_cs_control_wm_asset.erl (hunk 2 above)
Comment (cmeiklejohn) [inline src/riak_cs_control_wm_asset.erl]: Addressed in [ca1294531cbf841644f7b60b1eb2ddf60a44e677].

**Comment by cmeiklejohn** [comment]
I'm working on debugging the erlydtl errors now, but the 403 not updating error is blocked on us upgrading to the latest ember-data which I'll handle in a subsequent pull request.

**Comment by cmeiklejohn** [comment]
@reiddraper which version of Erlang are you running?  I just built a brand new rel, against erlydtl master with R15B03 and I'm not getting that CSRF error.

**Comment by cmeiklejohn** [comment]
Opened #20 to track the error handling for both user creation, and user modification operations.

**Comment by reiddraper** [comment]
> which version of Erlang are you running?

R15B01

**Comment by cmeiklejohn** [comment]
@reiddraper I can't reproduce this with R15B01 either.  Can you try rebuilding from source and verity this is still an issue?

**Comment by reiddraper** [comment]
wfm now, +1

**Comment by cmeiklejohn** [comment]
merging.

**Pull request merged**